cli/src/diff_util.rs

```diff
@@ -346,19 +347,65 @@ fn show_color_words_diff_line(
     Ok(())
 }
 
-fn diff_content(path: &RepoPath, value: MaterializedTreeValue) -> Result<Vec<u8>, CommandError> {
+struct FileContent {
+    /// true if this file is likely text; false if it is definitely binary.
+    is_binary: bool,
+    // TODO: maybe this could store an `enum { String, Vec<u8> }` instead? but then we have to
+    // UTF-8 validate the whole file.
+    contents: Vec<u8>,
+}
+
+impl FileContent {
+    pub(crate) fn is_empty(&self) -> bool {
+        self.contents.is_empty()
+    }
+}
+
+impl From<String> for FileContent {
+    fn from(contents: String) -> Self {
+        FileContent {
+            is_binary: false,
+            contents: contents.into_bytes(),
+        }
+    }
+}
+
+fn diff_file_content(reader: &mut dyn io::Read) -> io::Result<FileContent> {
+    // If this is a binary file, don't show the full contents.
+    // Determine whether it's binary by whether the first 8k bytes contain a null
+    // character; this is the same heuristic used by git as of writing: https://github.com/git/git/blob/eea0e59ffbed6e33d171ace5be13cde9faa41639/xdiff-interface.c#L192-L198
```

Comment:
It looks like you only check whether the file is UTF8, and not whether it contains the 0 byte. I thought the 0 byte is valid UTF8.

I might be wrong on either point, but perhaps add a test where there's a file that contains ASCII and the 0 byte?

Reply:
oh fantastic catch, thank you. the 0 byte is indeed valid utf8.